docs/style.rst

```diff
@@ -36,51 +36,154 @@ We like to use `>&` for file descriptors (numbers), and `&>` for filenames
+     some_command ${optional_flag} foo bar # noquotes
+     # The preferred way is to use arrays, if it is not overly cumbersome.
+     ${DRYRUN[@]+"${DRYRUN[@]}"} some command
+     some_command ${optional_flag[@]+"${optional_flag[@]}"} foo bar
+     # Since arrays cannot be exported, this is often not viable.
+
+     # Another reason quotes must not be used, is when splitting up a string into an array
+     foo="aa:bb:cc:dd"
+     IFS=":"
+     # This is actively splitting apart a string, and must not be in quotes
+     bar=(${foo}) # noquotes
+
+  Quotes should actually not be used in ``[[]]`` expressions. There are a few corner cases the will be treated literally. ``# noquotes`` is not needed for ``[[]]`` expressions.
+
+  .. code-block:: bash
 
-    echo ${avar} # WRONG style. It is safer to always quote the variable,
-                 # even if it is not strictly necessary
+     if [ "${var}" -gt "0" ] && [[ ${foo} =~ ${pattern} ]]; then
+       echo "hi"
+     fi
 
-* Prefer ``[ ]`` tests to the ``[[ ]]`` construct and = to ==
+  Simple assignments _may_ skip quotes, ``# noquotes`` is not needed as this is easy enough for a code reviewer to see
 
   .. code-block:: bash
 
-    [ "${avar}" = "foo bar" ]  # Variables are always quoted in [] tests
+     local a=1
+     b=2
+     x=(11 22 33 44 "5 5" "6  6")
+     cvar=foo
+     dvar=foo\ bar  # Wrong style. This is no longer simple. Use quotes
+     dvar="foo bar" # Right
 
-    [[ "${avar}" == "foo bar" ]] # WRONG style. Use []
+  It is best not to use quotes when inside of ``{}``. The expressions inside of ``{}`` can be thought of as already being implicitly quoted (``"``). Adding quotes (``'`` or ``"``) may seem to work at first, but the behavior of explicitly quoting will change between the different versions of bash.
 
-    [[ ${avar} = foobar* ]] # Ok. Pattern matching is not possible with []
+  .. code-block:: bash
 
-    [[ ${avar} = "foo bar"* ]] # WRONG style. If quotes are needed, use a variable
-    pattern="foo bar*"
-    [[ ${avar} = ${pattern} ]] # Ok. Also, never quote variables in [[ ]] as
-                               # this disables pattern matching---in which case,
-                               # [] can be used instead
+     echo "${foo-"b a  r"}"        # Wrong style
+     echo "${foo-b a  r}"          # Right
+     echo "${foo/"o"/"O"}"         # Wrong style
+     echo "${foo/o/O}"             # Right
+     echo "${foo/"  "/two spaces}" # Wrong style
+     echo "${foo/  /two spaces}"   # Right
+
+  :var:`bash_behavior_pattern_substitution_slash_escape_with_single_quote` is a special case that still needs special care, due to differences in bash behavior between versions.
+
+* Always use ${var} vs $var
+
+  The reason for this policy is consistency and to clarify that certain features in bash only work in the ``{}``, e.g. variable substitution. It's very easy for someone to mistake ``${foo+set}`` for ``$foo+set`` and not ``${foo}+set``.
```

Comment:
is this backwards? should it be, "It's very easy for someone to mistake ``$foo+set`` for ``${foo}+set`` and not ``${foo+set}`` or something?